src/annotator/anchoring/trim-range.ts

```diff
@@ -0,0 +1,225 @@
+  baseOffset: number,
+  direction: TrimDirection
+): number {
+  const nextChar =
+    direction === TrimDirection.Forwards ? baseOffset : baseOffset - 1;
+  if (text.charAt(nextChar).trim() !== '') {
+    // baseOffset is already a valid offset
+    return baseOffset;
+  }
+
+  let availableChars: string;
+  let availableNonWhitespaceChars: string;
+
+  if (direction === TrimDirection.Backwards) {
+    availableChars = text.substring(0, baseOffset);
+    availableNonWhitespaceChars = availableChars.trimEnd();
+  } else {
+    availableChars = text.substring(baseOffset);
+    availableNonWhitespaceChars = availableChars.trimStart();
+  }
+
+  if (!availableNonWhitespaceChars.length) {
+    return -1;
+  }
+
+  const offsetDelta =
+    availableChars.length - availableNonWhitespaceChars.length;
+
+  return direction === TrimDirection.Backwards
+    ? baseOffset - offsetDelta
+    : baseOffset + offsetDelta;
+}
+
+/**
+ * Calculate a new Range start position (TrimDirection.Forwards) or end position
+ * (Backwards) for `range` that represents the nearest non-whitespace character,
+ * moving into the `range` away from the relevant initial boundary node towards
+ * the terminating boundary node.
+ *
+ * The search is non-inclusive of the initial boundary node, but does consider
+ * the terminating boundary node.
+ *
+ * @throws {RangeError} If no text node with non-whitespace characters found
+ */
+function closestNonSpacePosition(
+  range: Range,
+  direction: TrimDirection
+): RangePosition {
+  const nodeIter =
+    range.commonAncestorContainer.ownerDocument!.createNodeIterator(
+      range.commonAncestorContainer,
+      NodeFilter.SHOW_TEXT
+    );
+
+  const initialBoundaryNode =
+    direction === TrimDirection.Forwards
+      ? range.startContainer
+      : range.endContainer;
+
+  const terminalBoundaryNode =
+    direction === TrimDirection.Forwards
+      ? range.endContainer
+      : range.startContainer;
+
+  let currentNode = nodeIter.nextNode();
+
+  // Advance the NodeIterator to the `initialBoundaryNode`
+  while (currentNode && currentNode !== initialBoundaryNode) {
+    currentNode = nodeIter.nextNode();
+  }
+
+  if (direction === TrimDirection.Backwards) {
+    // Reverse the NodeIterator direction. This will return the same node
+    // as the previous `nextNode()` call (initial boundary node).
+    currentNode = nodeIter.previousNode();
+  }
+
+  let trimmedOffset = -1;
+
+  const advance = () => {
+    currentNode =
+      direction === TrimDirection.Forwards
+        ? nodeIter.nextNode()
+        : nodeIter.previousNode();
+
+    if (currentNode) {
+      const nodeText = currentNode.textContent!;
+      const baseOffset =
+        direction === TrimDirection.Forwards ? 0 : nodeText.length;
+      trimmedOffset = closestNonSpaceOffset(nodeText, baseOffset, direction);
+    }
+  };
+
+  // Start the examination at the first text node before/after the
+  // initial boundary node, i.e. do not evaluate the `container` itself
```

Comment:
Re. the "i.e. ..." at the end, there is no `container` variable here. To restate what is going on, this function is called when the caller knows that the `startContainer` or `endContainer` (depending on direction) is not a suitable start/end point for a trimmed range, and so wants to advance to the next point?

I note that the tests still pass if you remove this `advance` call, so in fact it is just an optimization - unless there is some test case you can think of that would show a difference in behavior? You could simplify the function's behavior conceptually by removing it.

Reply:
All righty, I've removed it, and also updated the function's comment as the initial boundary node now _is_ considered in the search.

It's a non-meaningful change relative to this module because this function (newly-named `closestNonSpaceInRange`) is only called by the exported `trimRange` function if it was unable to find a valid offset in the original bounding node via `closestNonSpaceInString`. That is, by the time this function is called, we already "know" that the initial boundary node isn't valid, so there is no behavior change by not `advance()`-ing past it explicitly. If this function were part of the public API, it would represent a behavioral change.

It would be possible to do away with the initial calls to `closestNonSpaceInString` from the `trimRange` function and use only this function (now that it considers boundary nodes). However, using `closestNonSpaceInString` doesn't require node traversal and should be faster.